File: 0773-sliding-puzzle/0773-sliding-puzzle.java
class Solution {
    public int slidingPuzzle(int[][] board) {
        int r = board.length;
        int c = board[0].length;

        StringBuilder source = new StringBuilder();
        String target = "123450";

        for (int i = 0; i < r; i++) {
            for (int j = 0; j < c; j++) {
                source.append(board[i][j]);
            }
        }

        HashMap<Integer, List<Integer>> m = new HashMap<>();
        m.put(0, Arrays.asList(1, 3));
        m.put(1, Arrays.asList(0, 2, 4));
        m.put(2, Arrays.asList(1, 5));
        m.put(3, Arrays.asList(0, 4));
        m.put(4, Arrays.asList(1, 3, 5));
        m.put(5, Arrays.asList(2, 4));

        HashMap<String, Integer> vis = new HashMap<>();
        Queue<Pair<String, Integer>> q = new LinkedList<>();

        vis.put(source.toString(), 1);
        q.offer(new Pair<>(source.toString(), 0));

        int ans = Integer.MAX_VALUE;

        while (!q.isEmpty()) {
            String s = q.peek().getKey();
            int lvl = q.peek().getValue();
            q.poll();

            if (s.equals(target)) {
                ans = Math.min(ans, lvl);
            }

            int ind = -1;
            for (int i = 0; i < s.length(); i++) {
                if (s.charAt(i) == '0') {
                    ind = i;
                    break;
                }
            }

            for (int x : m.get(ind)) {
                StringBuilder temp = new StringBuilder(s);
                char ch = temp.charAt(x);
                temp.setCharAt(x, '0');
                temp.setCharAt(ind, ch);

                if (!vis.containsKey(temp.toString())) {
                    vis.put(temp.toString(), 1);
                    q.offer(new Pair<>(temp.toString(), lvl + 1));
                }
            }
        }

        return (ans == Integer.MAX_VALUE) ? -1 : ans;
    }
}

File: 0773-sliding-puzzle/0773-sliding-puzzle.cpp
class Solution {
public:
    int slidingPuzzle(vector<vector<int>>& board) {
        int r = board.size();
        int c = board[0].size();
        
        string source;
        string target = "123450";
        for(int i=0;i<r;i++){
            for(int j=0;j<c;j++){
                source += to_string(board[i][j]);
            }
        }
        
        unordered_map<int,vector<int>> m;  // 0 kis kis se swap mar sakta hai
        m[0] = {1,3}; m[1] = {0,2,4}; m[2] = {1,5};
        m[3] = {0,4}; m[4] = {1,3,5}; m[5] = {2,4};
        
        
        unordered_map<string,int> vis;
        queue<pair<string,int>> q;
        
        vis[source] = 1;
        q.push({source,0});
        
        int ans = INT_MAX;
        while(!q.empty()){
            string s = q.front().first;
            int lvl = q.front().second;
            q.pop();
            if(s == target){
                ans = min(ans,lvl);
            }
            int ind = -1;
            for(int i=0;i<s.size();i++){
                if(s[i] == '0'){
                    ind = i;
                    break;
                }
            }
            for(auto x:m[ind]){
                string temp = s;
                swap(temp[ind],temp[x]);
                if(vis.find(temp) == vis.end()){
                    vis[temp] = 1;
                    q.push({temp,lvl+1});
                }
            }
        }
        if(ans == INT_MAX)
            return -1;
        return ans;
    }
};